### ISSUE TITLE
Documentations for custom quantization methods integration are needed

### ISSUE THREAD

**Issue by Kelvinyu1117** [Xilinx/brevitas#226]
Are there any documentations or examples for integrating my custom quantization algorithm into the existing layer modules?

**Comment by volcacius** [comment]
Hello,

Yeah sorry I'm very behind on writing documentation so there aren't really examples yet, but it's supported and I'm happy to help anyone interested. In general you can implement a custom quantizer the same way current quantizers are implemented. 

The first thing to know is that Brevitas does dependency injection using https://github.com/proofit404/dependencies . This way a quantizer can be instantiated by Brevitas without explicit knowledge of how its internal components relate to each other.

You define a quantizer by inheriting from the BaseInjector as done here https://github.com/Xilinx/brevitas/blob/aff49758ec445d77c75721c7de3091a2a1797ca8/brevitas/inject/defaults.py#L22 

BaseInjector is just a variation of the Injector defined by the 'dependencies' library that fixes issues with interactive debugging.
 The injector has to define a `tensor_quant` attribute that points to the implementation of your quantizer. Any argument that required for the initialization of the quantizer should be set in the injector. You can follow the documentation of the 'dependencies' library to understand the mechanism.

The implementation of your quantizer should accept a single argument (the value to be quantized) and return the quantized tensor(in a float32 tensor datatype), the output scale and bit width, such that the following relationship holds:
*y / scale* is an integer that can be represented with *bit_width* bits (there isn't support for zero_point yet, but I'm gonna add it over the next few days). All of them should be pytorch tensors.

 So say:

```python
class MyQuantizer(torch.nn.Module):
    def __init__(self, arg_one, arg_two):
        ...

    def forward(x: Tensor) -> Tensor, Tensor, Tensor:
        ... # x is the values to be quantized, y is the quantized value
        return y, scale, bit_width

MyQuantInjector(BaseInjector):
    tensor_quant = MyQuantizer   
    arg_one = whatever1
    arg_two = whatever2
```

Finally, you can pass the Injector to the various layers. For layers with weights and biases like QuantConv2d or QuantLinear, you should pass `weight_quant=MyQuantInjector` for quantization of weights. You can also quantize inputs and outputs by setting _input_quant_ and _output_quant_. For activation layers like QuantReLU, you can set _input_quant_ and _act_quant_ (which quantizes the output). You can quantize only the parts that are relevant to your use case, anything you don't need you can set to None.

In some scenarios though your quantizer may need during `__init__` information defined by the module you are quantizing, say for example the size of the weight tensor. You can then define a function that updates MyQuantInjector based on the module it's being passed to. The signature is:

```
update_my_quant_injector(layer: Module, injector: Injector,  prefix: str, **kwargs) -> Injector:
    injector = injector.let(weight_shape=module.weight.shape)  # example, pass shape of weights to injector
    return injector
```

You can then pass it to say a QuantConv2d by setting *update_wqi=update_my_quant_injector* (*update_iqi* for inputs, *update_oqi* for outputs). In that case *layer* would be the QuantConv2d object, *injector* is MyQuantInjector, *prefix* is a string that is 'weight', 'input' or 'output' depending on what you are setting, *kwargs* are all additional keyword arguments passed to QuantConv2d. If you don't need to perform any updates, you can just set update_wqi=None (same for the others). For activation layers like QuantReLU, you can set *update_iqi* and *update_aqi* to update respectively the input and output quantization injectors.
The default updater for weights is here. https://github.com/Xilinx/brevitas/blob/master/brevitas/inject/solver.py#L477

So to wrap up you have:

```python
class MyQuantizer(torch.nn.Module):
    def __init__(self, arg_one, arg_two, weight_shape):
        ...

    def forward(x: Tensor) -> Tensor, Tensor, Tensor:
        ... # x is the values to be quantized, y is the quantized value
        return y, scale, bit_width

MyQuantInjector(BaseInjector):
    tensor_quant = MyQuantizer   
    arg_one = whatever1
    arg_two = whatever2

update_my_quant_injector(layer: Module, injector: Injector,  prefix: str, **kwargs) -> Injector:
    injector = injector.let(weight_shape=layer.weight.shape)  # example, pass shape of weights to injector
    return injector

quant_conv = QuantConv2d(..., weight_quant=MyQuantInjector, update_wqi=update_my_quant_injector)
```

Brevitas itself has an additional layer of abstraction where instead of passing directly MyQuantizer to MyQuantInjector I'm passing an enum that depending on the context then gets resolved to a certain quantizer. This is just to provide a simpler user experience. The transformation is performed in the updater linked above.

**Comment by Kelvinyu1117** [comment]
Thank you for your reply! I am trying to implement my quantizer in your suggested way now.
But I got a question that how can I initialize my model using the pre-trained weights (full precision)? 

I want to start with a full precision model (should be fp32 by default in Pytorch) with pre-trained weight, then adding my quantizer and train the model in order to check how much accuracy can be maintained after the quantization applied.

One of the ways I can think of is setting both the `bit_width` and `weight_bit_width` to 32. But I am not sure is it the right way to do so.

**Comment by volcacius** [comment]
Hello,

Regarding loading pretrained weights, as long as your quantized model has the same structure of your floating point model (in terms of how modules are nested) you can simply do `quant_model.load_state_dict(float_state_dict)`. 
If you are using Brevitas built-in quantizers you might get an error regarding missing keys in the state dict. That's because some quantizers internally learn a `torch.nn.Parameter` (e.g. `scaling_impl_type='parameter'` or `bit_width_impl_type='parameter'`), but those parameters are obviously not to be found in the float_state_dict. To fix it, you can simply set the env variable `BREVITAS_IGNORE_MISSING_KEYS=1`.

Regarding disabling quantization, there are two ways. You can simply set the whole  `weight_quant` / `act_quant` / `input_quant` / `output_quant` injector to `None`, or you can set (input_/output_/weight_)`quant_type='fp'` . Either way you get that a QuantConv2d behaves just like a Conv2d. 

Regarding your comment on `bit_width`, if you set them to 32 you would get 32bit integer quantization (which is not a good idea on top of a float32 datatype). Also keep in mind that `bit_width` and `weight_bit_width` are all arguments defined by Brevitas' built-in quantizers. If you define your own quantizers, either you implement your own mechanism to disable quantization (like Brevitas' quantizers do when you set quant_type='fp'), or you set the injectors to `None` as I explained above.

Alessandro

**Comment by Kelvinyu1117** [comment]
Hi,

Thank you for your reply! I have implemented my quantizer with some learnable parameter based on your tutorial as follow:

```
class MyQuantizer(torch.nn.Module):
    def __init__(self, bit_width, is_weight):
         self.bit_width = bit_width
         self.s = torch.nn.Parameter(torch.ones(1)) # it is a learnable parameter (step size)
         self.is_weight = is_weight
         ...
     def forward(self, x: Tensor) -> (Tensor, Tensor, Tensor):  
         ... return 
class MyQuantizeWeightInjector8Bits(BaseInjector):
    tensor_quant = MyQuantizer
    bit_width = 8
    is_weight = True

class MyQuantizeInputInjector8Bits(BaseInjector):
    tensor_quant = MyQuantizer
    bit_width = 8
    is_weight = False

```
For the Linear Layer, Conv Layer and the activation, I use the following:

```
QuantConv2d(..., weight_quant=MyQuantizeWeightInjector8Bits, input_quant=MyQuantizeInputInjector8Bits,  output_quant=None)

QuantLinear(..., weight_quant=MyQuantizeWeightInjector8Bits, input_quant=MyQuantizeInputInjector8Bits, output_quant=None)

QuantReLU(
            input_quant=None, act_quant=None, output_quant=None)
```
I use the above approach to modify the example LeNet and test it on MINST dataset.

I have tested the example LeNet with Brevitas default quantizer against my quantizer (same network structure just changing the quantizer), and the example network works perfectly on the dataset but my network seems to learn nothing from the dataset. The loss and accuracy are very low and almost not changing during each epoch:


They are using the same training setup and code, so I don't it is because of the wrong implementation of the training code like the missing of `optimizer.zero_grad()`.

Before migrating the quantizer to Brevitas, I have implemented the quantizer and integrate to my model in pure Pytorch and it should be correctly implemented as nothing weird happened. And I just simply copy the `forward` logic from the old one to Brevitas and return the quantized output, bit width, step size, so it should not because of the wrong implementation of the model.

Am I missing something when I integrate the quantizer? 

Thank you

**Comment by volcacius** [comment]
Hello,

I would try a couple of things. First, I would set the update_*qi to None for any injector you are setting to != None, if there are name clashes it might be that the default one is interfering somehow (I the long term I would like to get rid of it, but at the moment it takes of a lot of retrocompatibility issues with previous versions of the library). 
After that, I'd try to quantize first only weights, then only activations, and then weights and activations together. You might be able to pinpoint which part is problematic. 
Other than that unfortunately without looking at the whole code it's hard to say what might be going on.

Alessandro

**Comment by Kelvinyu1117** [comment]
Hi,

I have set the update_*qi to None for all injector and check with different quantization schemes (quantize only the weights, quantize only the activations) that you mention above, the result is the following:

Quantize only the weights: the network parameters don't update, that's the parameter remains unchanged, return very low accuracy (~10%)

Quantize only the activations (the ReLU): the accuracy is making sense (~98%) on MNIST for the example LeNet. 

As I am using the same quantizer both test, which means the bound of the quantization are the same, I think the problem should be related to the misuses of the library.

**Comment by volcacius** [comment]
See https://github.com/Xilinx/brevitas/blob/master/notebooks/03_anatomy_of_a_quantizer.ipynb .
